Python/Product/Analysis/LanguageServer/Server.SignatureHelp.cs

```diff
@@ -0,0 +1,153 @@
+            if (activeParameter >= 0 && activeSignature < 0) {
+                // TODO: Better selection of active signature
+                activeSignature = sigs
+                    .Select((s, i) => Tuple.Create(s, i))
+                    .OrderBy(t => t.Item1.parameters.Length)
+                    .FirstOrDefault(t => t.Item1.parameters.Length > activeParameter)
+                    ?.Item2 ?? -1;
+            }
+
+            activeSignature = activeSignature >= 0
+                ? activeSignature
+                : (sigs.Length > 0 ? 0 : -1);
+
+            var sh = new SignatureHelp {
+                signatures = sigs,
+                activeSignature = activeSignature,
+                activeParameter = activeParameter
+            };
+            return sh;
+        }
+
+        private SignatureInformation ToSignatureInformation(IOverloadResult overload) {
+            var si = new SignatureInformation();
+
+            if (_clientCaps?.textDocument?.signatureHelp?.signatureInformation?._shortLabel ?? false) {
+                si.label = overload.Name;
+            } else {
+                si.label = "{0}({1})".FormatInvariant(
+                    overload.Name,
+                    string.Join(", ", overload.Parameters.Select(FormatParameter))
+                );
+            }
+
+            si.documentation = string.IsNullOrEmpty(overload.Documentation) ? null : overload.Documentation;
+            si.parameters = overload.Parameters.MaybeEnumerate().Select(p => new ParameterInformation {
+                label = p.Name,
+                documentation = string.IsNullOrEmpty(p.Documentation) ? null : p.Documentation,
+                _type = p.Type,
+                _defaultValue = p.DefaultValue
+            }).ToArray();
+
+            switch (SelectBestMarkup(_clientCaps.textDocument?.signatureHelp?.signatureInformation?.documentationFormat, MarkupKind.Markdown, MarkupKind.PlainText)) {
+                case MarkupKind.Markdown:
+                    var converter = new RestTextConverter();
+                    if (!string.IsNullOrEmpty(si.documentation.value)) {
+                        si.documentation.kind = MarkupKind.Markdown;
+                        si.documentation.value = converter.ToMarkdown(si.documentation.value);
+                    }
+                    foreach (var p in si.parameters) {
+                        if (!string.IsNullOrEmpty(p.documentation.value)) {
+                            p.documentation.kind = MarkupKind.Markdown;
+                            p.documentation.value = converter.ToMarkdown(p.documentation.value);
+                        }
+                    }
+                    break;
+            }
+
+            si._returnTypes = (overload as IOverloadResult2)?.ReturnType.OrderBy(k => k).ToArray();
+            return si;
+        }
+
+        private MarkupKind SelectBestMarkup(IEnumerable<MarkupKind> requested, params MarkupKind[] supported) {
```

Comment:
Seems like this will be used from a few places - should we have a "Server.PrivateHelpers.cs" partial class as well?